### PR TITLE
Adding https support to ServiceBuilder

### PR DESCRIPTION
**Pull request by ZaradarBH**
#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [x] Screenshots attached (for UI changes)
- [x] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [x] Regression tests added for bug fixes

### CODE REVIEW

**Comment by ZaradarDFDS** [comment]
See comment on https://github.com/spotify/backstage/pull/1738 for historical reference.

**Comment by Rugvip** [comment]
Thanks!

Continued some threads in #1738, but please let us know when this is ready for another review and we'll move the discussions over.

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts [hunk 1]
```diff
@@ -120,34 +154,96 @@ export class ServiceBuilderImpl implements ServiceBuilder {
...
+      if (httpsSettings) {
+        logger.info('Initializing https server');
+
+        let credentials: { key: string; cert: string };
+        const signingOptions: any = httpsSettings?.certificate;
+
+        if (signingOptions?.algorithm !== undefined) {
+          const certificateAttributes: Array<any> = [];
+
+          Object.getOwnPropertyNames(signingOptions?.attributes).forEach(
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:168]: You can use `Object.entries(signingOptions.attributes).map(([name, value]) => ({ name, value }))` here instead.

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts [hunk 2]
```diff
@@ -120,34 +154,96 @@ export class ServiceBuilderImpl implements ServiceBuilder {
...
+              algorithm: signingOptions?.algorithm || 'sha256',
+              days: signingOptions?.days || 30,
+            },
+          );
+
+          credentials = { key: signatures.private, cert: signatures.cert };
+        } else {
+          credentials = { key: '', cert: '' };
+
+          const keyPath = path.resolve(__dirname, signingOptions?.key);
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:195]: I think the expected behaviour here is that the certificate path is relative to the config file, so imo it's better to rely on the config secrets to load in certificates from files instead. Using the example in the previous PR you'd just get the entire certificate file loaded into the config `key` and `cert` values.

File: packages/backend-common/src/service/lib/config.ts [hunk 3]
```diff
@@ -36,12 +57,14 @@ export type BaseOptions = {
  * }
  * ```
  */
-export function readBaseOptions(config: ConfigReader): BaseOptions {
+export function readBaseOptions(configReader: ConfigReader): BaseOptions {
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/config.ts:60]: Let's keep the config name here

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts [hunk 4]
```diff
@@ -120,34 +154,96 @@ export class ServiceBuilderImpl implements ServiceBuilder {
...
+
+              certificateAttributes.push({
+                name: propertyName,
+                value: propertyValue,
+              });
+            },
+          );
+
+          // TODO: Create a type def for selfsigned.
+          const signatures = require('selfsigned').generate(
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:182]: I'd say it's worth logging before this step, something like "generating self-signed certificate with attributes ...". Just to provide a bit of information and show what's going on, as this step can take a while.

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts [hunk 5]
```diff
@@ -30,16 +32,25 @@ import {
...
+  readBaseOptions,
+  readCorsOptions,
+  readHttpsSettings,
+  HttpsSettings,
+} from './config';
 
+const DEFAULT_PROTOCOL = 'https://';
 const DEFAULT_PORT = 7000;
 const DEFAULT_HOST = 'localhost';
+const DEFAULT_BASEURL = `${DEFAULT_PROTOCOL}${DEFAULT_HOST}:${DEFAULT_PORT}`;
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:45]: We can get rid of this and the baseUrl option itself now right, doesn't seem to be used?

File: packages/backend-common/src/service/lib/config.ts [hunk 6]
```diff
@@ -20,6 +20,27 @@ import { CorsOptions } from 'cors';
...
   listenHost?: string;
+  baseUrl?: string;
+};
+
+export type HttpsSettings = {
+  certificate?: CertificateSigningOptions | CertificateFileOptions;
+};
+
+export type CertificateFileOptions = {
+  key?: string;
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/config.ts:31]: These should probably both be required right?

File: packages/cli/src/commands/create-app/createApp.ts [hunk 7]
```diff
@@ -112,10 +113,11 @@ export default async (cmd: Command): Promise<void> => {
       choices: ['PostgreSQL', 'SQLite'],
     },
   ];
-  const answers: Answers = await inquirer.prompt(questions);
+  
+  const answers: Answers = await mapInqueryAnswersFromCommanderOptions(questions, cmd);
```
Comment (Rugvip) [inline packages/cli/src/commands/create-app/createApp.ts:117]: Could we do this in a separate PR? 😅

Comment (ZaradarDFDS) [reply packages/cli/src/commands/create-app/createApp.ts:117]: It seems as if whenever we make a PR to our own dfds:master it gets auto-merged into any open PR from dfds:master to spotify:master. Any idea on what is causing this behavior?

Comment (Rugvip) [reply packages/cli/src/commands/create-app/createApp.ts:117]: Yeah it's intended behaviour, PRs are not snapshots but a living diff of merging one branch into another. When you merge a PR towards master in your fork, that gets reflected in this PR, as it's from the master branch of your fork.

One way of opening the PR in this case would be to keep a feature branch around in your fork, and make the PR from that. If it's all already merged into your master, you could create a new branch and bring in only the changes that you want to create a PR from. Some more info here and in the surrounding docs: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork

Comment (ZaradarDFDS) [reply packages/cli/src/commands/create-app/createApp.ts:117]: Thanks!

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts [hunk 8]
```diff
@@ -120,20 +142,95 @@ export class ServiceBuilderImpl implements ServiceBuilder {
...
+          );
+
+          credentials.key = signatures.private;
+          credentials.cert = signatures.cert;
+        } else {
+          if (fs.existsSync(signingOptions?.key)) {
+            if (fs.lstatSync(signingOptions?.key).isFile()) {
+              logger.info('Bootstrapping key from file');
+
+              credentials.key = fs.readFileSync(signingOptions?.key).toString();
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:189]: This will load the key file relative to the working directory, which can be quite unpredictable. For example, an `app-config.yaml` in the root of the repo that specifies `key: https-key.pem`, would actually refer to the path `'packages/backend/https-key.pem'` if you're running the backend as usual.

I think it's better if the path is readable and makes sense as viewed in the `app-config.yaml`, and that paths are relative to the config files themselves. The clean way to do that atm is to just let the full key be configured, and pull in the files using the build-in secret support in the config, e.g. `$secret: file: https-key.pem`

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts [hunk 9]
```diff
@@ -120,20 +142,95 @@ export class ServiceBuilderImpl implements ServiceBuilder {
...
+            certificateAttributes,
+            {
+              keySize: signingOptions?.size || 2048,
+              algorithm: signingOptions?.algorithm || 'sha256',
+              days: signingOptions?.days || 30,
+            },
+          );
+
+          logger.info(
+            'Bootstrapping key and cert from self-signed certificate',
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:179]: Was thinking logging before the `generate()` call, at this point we're already done waiting 😅

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts [hunk 10]
```diff
@@ -120,20 +142,95 @@ export class ServiceBuilderImpl implements ServiceBuilder {
...
+          cert: '',
+        };
+        const signingOptions: any = httpsSettings?.certificate;
+
+        if (signingOptions?.algorithm !== undefined) {
+          logger.info('Generating self-signed certificate with attributes');
+
+          const certificateAttributes: Array<any> = [];
+
+          Object.entries(signingOptions?.attributes).map(([name, value]) => {
```
Comment (Rugvip) [inline packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:161]: Better use `forEach` or a `for (... of ...)` loop when you're not using the return value of the iterator function.

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts (hunk 8 above)
Comment (ZaradarDFDS) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:189]: To be honest if people enter a relative file URL they should be aware that it will be relative to the host, unless they specify a absolute URL in which case it should resolve if the host has access to it. Regardless of what we make the "relative root" it will always confuse people, some will expect it to be the "logical root" (plugin root folder), others will expect it to be the mono repo root... However in my view forcing a standalone node module to rely on anything other then its "logical root" could create coupling between the "core system" and the "plugin system" which could potentially make it harder distributed the backend further down the line?

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts (hunk 9 above)
Comment (ZaradarDFDS) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:179]: I measured the time it take to generate a 2048 bit sha256 on my 8 year old home PC with an intel 7 chip. Its in the millisecond range so I dont think it matters to much and for me its more important to inform people that https was bootstrapped using whatever strategy they configured. With that said you cant have to many logging calls so I will add another one. :)

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts (hunk 10 above)
Comment (ZaradarDFDS) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:161]: You asked me to change it to this variant two days ago (check resolved comments). Do you want me to replace the .map with a forEach? Its not a problem, but I figured I would just use your suggestions AS-IS? :)

Comment (Rugvip) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:161]: `Object.entries(signingOptions.attributes).map(([name, value]) => ({ name, value }))` is what I suggested, plain functional style. Imperative style works too but it's a bit cleaner to not use `map` in that case.

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts (hunk 9 above)
Comment (Rugvip) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:179]: Yeah saw you added both now so it's fine

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts (hunk 8 above)
Comment (Rugvip) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:189]: Resolving it relative to the config file completely removes coupling to the packages, and makes it very portable. Imo it's less surprising and you can know what your config will be by just looking at the repo source.

**Review by Rugvip — COMMENTED** [review]
Gonna need a new PR that isn't from your master 😅

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts (hunk 8 above)
Comment (ZaradarDFDS) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:189]: It depends on how you look at it I guess. I think the plugins, more specifically standalone processes, should be fully self-contained as this will ease the process of "chopping up the backend". This would probably require our cfg system to support nested cfgs, chopping up our root config and tweeking some cfg loaders. But I think the end result would make it easier for non-devs to identify the different logical processes that make up the application and deploy them independently without bringing everything and the kitchen sink :p

Comment (Rugvip) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:189]: Config would still be a part of that chopped up backend, so it's not really related to what we're discussing

File: packages/backend-common/src/service/lib/ServiceBuilderImpl.ts (hunk 10 above)
Comment (ZaradarDFDS) [reply packages/backend-common/src/service/lib/ServiceBuilderImpl.ts:161]: With my limited understanding of NodeJS/TS it seems to me like the original suggestion did not really take into account that we needed to push the mapped value to the certAttributes array. Which is why I augmented the map function to handle this. Now if where is a cleaner way to do it, then what I just implemented, I have no problem refactoring it to meet that requirement. However as it stands I cannot revert to your AS-IS suggestion without modification because in my understanding that stop the propagation of values to the certAttr array and break the self-signed signature generation.

**Comment by ZaradarDFDS** [comment]
I will close the PR. Our discovery phase is ending and based on our assessment after I am done with my one week of vacation we will re-open new PRs if we decide to continue on this path. :)
